service/status-report/main.go

```diff
@@ -138,10 +141,46 @@ func main() {
 	status := providerclient.NewStorageClientStatus().
 		SetPlatformVersion(pltVersion).
 		SetOperatorVersion(oprVersion)
-	if _, err = providerClient.ReportStatus(ctx, storageClient.Status.ConsumerID, status); err != nil {
+	statusResponse, err := providerClient.ReportStatus(ctx, storageClient.Status.ConsumerID, status)
+	if err != nil {
 		klog.Exitf("Failed to report status of storageClient %v: %v", storageClient.Status.ConsumerID, err)
 	}
 
+	versionFromResponse, err := getVersionFromChannel(statusResponse.DesiredClientOperatorChannel)
+	if err != nil {
+		klog.Exitf("Failed to map channel name from status response to version: %v", err)
+	}
+
+	var updateAnnotation bool
+	if storageClient.GetAnnotations()[desiredSubscriptionChannelAnnotationKey] == "" {
+		updateAnnotation = true
+	}
+
+	var versionFromAnnotation *version.Version
+	if !updateAnnotation {
+		var err error
+		versionFromAnnotation, err = getVersionFromChannel(storageClient.GetAnnotations()[desiredSubscriptionChannelAnnotationKey])
+		if err != nil {
+			klog.Exitf("Failed to map channel name from storageclient annotation to version: %v", err)
+		}
+		// bump the annotation if it is less than version value from response, having higher version in annotation implies user overrode the upgrade
+		// negation is used here for checking strictly greater version
+		if !versionFromResponse.LessThan(versionFromAnnotation) {
+			// since we are referring only version from channel, by checkig for strictly greater version we'll enforce the channel name is also same
+			updateAnnotation = true
+		}
+	}
```

Comment:
[nb-ohad](https://github.com/nb-ohad) [Mar 15, 2024](https://github.com/red-hat-storage/ocs-client-operator/pull/89#discussion_r1525441807)

I am not sure why we are doing all of this. Channel names are assumed to have lexical ordering. You should be able to just compare the strings.


[leelavg](https://github.com/leelavg) [Mar 15, 2024](https://github.com/red-hat-storage/ocs-client-operator/pull/89#discussion_r1525844611)

I don't think so
```
# rlwrap -A yaegi
> // Compare (a, b string) int ==> 0 for a==b, -1 for a<b, +1 for a>b
> strings.Compare("stable-4.10", "stable-4.9")
: -1
>
```